### PR TITLE
Level properties page

### PR DESCRIPTION
**Pull request by shaobinspire**

### CODE REVIEW

File: Applications/Spire/Source/BookView/BookViewLevelPropertiesPage.cpp [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+      static const auto value = QObject::tr("Solid");
+      return value;
+    }
+  }
+
+  void scale(std::shared_ptr<ListModel<QColor>>& scheme, const QColor& start,
+      const QColor& end, int levels) {
+    auto colors = scale_oklch(start, end, levels);
+    auto alphas = scale_alpha(start.alpha(), end.alpha(), levels);
+    scheme->transact([&] {
```
Comment (kranar) [inline Applications/Spire/Source/BookView/BookViewLevelPropertiesPage.cpp:56]: Where is ownership of ```scheme``` shared?

File: Applications/Spire/Source/BookView/BookViewLevelPropertiesPage.cpp [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+      font.setWeight(QFont::Medium);
+      font.setPixelSize(scale_width(12));
+      style.get(Any()).
+        set(Font(font)).
+        set(PaddingBottom(scale_height(8)));
+    });
+    return label;
+  }
+
+  auto make_levels_slot(std::shared_ptr<OptionalIntegerModel> model) {
```
Comment (kranar) [inline Applications/Spire/Source/BookView/BookViewLevelPropertiesPage.cpp:92]: Don't call a parameter ```model``` anymore than you would call a variable ```variable```.

File: Applications/Spire/Source/BookView/BookViewLevelPropertiesPage.cpp [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+  QHBoxLayout* m_body_layout;
+  QVBoxLayout* m_color_bands_layout;
+  QVBoxLayout* m_color_boxes_layout;
+  std::array<CheckBox*, 2> m_buttons;
+  int m_gradient_color_levels;
+  scoped_connection m_levels_connection;
+  scoped_connection m_scheme_connection;
+  scoped_connection m_color_connection;
+
+  PriceLevelWidget( std::shared_ptr<ValueModel<std::vector<QColor>>> source,
```
Comment (kranar) [inline Applications/Spire/Source/BookView/BookViewLevelPropertiesPage.cpp:369]: Remove whitespace.

**Review by kranar — CHANGES_REQUESTED** [review]

File: Applications/Spire/Include/Spire/Ui/Ui.hpp [hunk 4]
```diff
@@ -163,6 +163,15 @@ namespace Spire {
...
   void invalidate_descendant_layouts(QWidget& widget);
+
+  /**
+   * Gets the width of a character.
+   * @param font The font used to calculate the character width.
+   */
+  int get_character_width(const QFont& font);
+
+  /** Gets the default width of a character in Spire. */
+  int get_default_character_width();
```
[the remark this answers is not in the document]
Comment (shaobinspire) [reply Applications/Spire/Include/Spire/Ui/Ui.hpp:174]: This function returns the character width using the default font of the `TextBox`, which is "Roboto" at 12px with a normal weight.

[the remark this answers is not in the document]
Comment (kranar) [reply Applications/Spire/Include/Spire/Ui/Ui.hpp:174]: Remove it and use only the ```get_character_width``` function.

**Review by kranar — APPROVED** [review]

**Pull request merged**